# 北上广深租房市场数据分析

## 第四部分：数据探索分析

In [1]:
from pathlib import Path
import pandas as pd

DATA_PATH = Path("../data/processed/rent_cleaned.csv")

df = pd.read_csv(DATA_PATH)

print("数据规模：", df.shape)

数据规模： (11978, 26)


## 一、城市租金比较

In [2]:
# 整租和合租的面积含义不同，因此分别建立分析数据
entire_df = df[
    df["type"] == "整租"
].copy()

shared_df = df[
    df["type"] == "合租"
].copy()

In [3]:
# 分别计算每个城市的整租房源数量、平均租金和租金中位数
entire_city_count = entire_df.groupby(
    "city"
)["rent_price_num"].count()

entire_city_mean = entire_df.groupby(
    "city"
)["rent_price_num"].mean().round(2)

entire_city_median = entire_df.groupby(
    "city"
)["rent_price_num"].median()

In [4]:
# 把三项统计结果组合成一张表
entire_city_rent = pd.DataFrame({
    "整租房源数量": entire_city_count,
    "整租平均月租金": entire_city_mean,
    "整租月租金中位数": entire_city_median
})

In [5]:
# 按照租金中位数从高到低排列
entire_city_rent = entire_city_rent.sort_values(
    "整租月租金中位数",
    ascending=False
)

entire_city_rent

,整租房源数量,整租平均月租金,整租月租金中位数
city,,,
北京,2992,8969.08,6500.0
上海,2992,9364.25,6000.0
深圳,2682,6661.09,4500.0
广州,2898,3943.05,3100.0


In [6]:
# 单独计算每个城市的合租房源数量、平均租金和租金中位数
shared_city_count = shared_df.groupby(
    "city"
)["rent_price_num"].count()

shared_city_mean = shared_df.groupby(
    "city"
)["rent_price_num"].mean().round(2)

shared_city_median = shared_df.groupby(
    "city"
)["rent_price_num"].median()

In [7]:
# 把三项统计结果组合成一张表
shared_city_rent = pd.DataFrame({
    "合租房源数量": shared_city_count,
    "合租平均月租金": shared_city_mean,
    "合租月租金中位数": shared_city_median
})

In [8]:
# 按照租金中位数从高到低排列
shared_city_rent = shared_city_rent.sort_values(
    "合租月租金中位数",
    ascending=False
)

shared_city_rent
# 发现北京无合租房源

,合租房源数量,合租平均月租金,合租月租金中位数
city,,,
上海,3,2566.67,2500.0
深圳,317,2161.79,2200.0
广州,94,1894.60,1845.0


In [9]:
# # 统一检查每个城市的整租、合租和总房源数量
city_type_count = pd.crosstab(
    df["city"],
    df["type"],
    margins=True,
    margins_name="总计"
)

city_type_count

type,合租,整租,总计
city,,,
上海,3,2992,2995
北京,0,2992,2992
广州,94,2898,2992
深圳,317,2682,2999
总计,414,11564,11978


## 二、城市面积比较

In [10]:
# 分别计算每个城市整租房源的平均面积和面积中位数
entire_city_area_mean = entire_df.groupby(
    "city"
)["rent_area_num"].mean().round(2)

entire_city_area_median = entire_df.groupby(
    "city"
)["rent_area_num"].median()

In [11]:
# 把两项统计结果组合成一张表
entire_city_area = pd.DataFrame({
    "整租平均面积": entire_city_area_mean,
    "整租面积中位数": entire_city_area_median
})

In [12]:
# 按照面积中位数从大到小排列
entire_city_area = entire_city_area.sort_values(
    "整租面积中位数",
    ascending=False
)

entire_city_area

,整租平均面积,整租面积中位数
city,,
上海,106.24,90.0
北京,91.34,78.0
广州,77.17,73.0
深圳,67.96,54.0


In [13]:
# 单独计算有合租样本城市的平均面积和面积中位数
shared_city_area_mean = shared_df.groupby(
    "city"
)["rent_area_num"].mean().round(2)

shared_city_area_median = shared_df.groupby(
    "city"
)["rent_area_num"].median()

In [14]:
# 把两项统计结果组合成一张表
shared_city_area = pd.DataFrame({
    "合租平均面积": shared_city_area_mean,
    "合租面积中位数": shared_city_area_median
})

In [15]:
# 按照面积中位数从大到小排列
shared_city_area = shared_city_area.sort_values(
    "合租面积中位数",
    ascending=False
)

shared_city_area

,合租平均面积,合租面积中位数
city,,
上海,75.33,65.0
广州,15.04,12.0
深圳,15.21,12.0


In [16]:
# 统一整理四个城市的整租和合租面积中位数
city_area_comparison = pd.DataFrame({
    "整租面积中位数": entire_city_area_median,
    "合租面积中位数": shared_city_area_median
})

In [17]:
# 按照整租面积中位数从大到小排列
city_area_comparison = city_area_comparison.sort_values(
    "整租面积中位数",
    ascending=False
)

city_area_comparison

,整租面积中位数,合租面积中位数
city,,
上海,90.0,65.0
北京,78.0,NaN
广州,73.0,12.0
深圳,54.0,12.0


## 三、城市每平方米租金比较

In [18]:
# 计算每个城市整租房源的平均每平方米租金和中位数
entire_city_per_sqm_mean = entire_df.groupby(
    "city"
)["rent_price_per_sqm"].mean().round(2)

entire_city_per_sqm_median = entire_df.groupby(
    "city"
)["rent_price_per_sqm"].median().round(2)

In [19]:
# 把两项统计结果组合成一张表
entire_city_per_sqm = pd.DataFrame({
    "整租平均每平方米租金": entire_city_per_sqm_mean,
    "整租每平方米租金中位数": entire_city_per_sqm_median
})

In [20]:
# 按照每平方米租金中位数从高到低排列
entire_city_per_sqm = entire_city_per_sqm.sort_values(
    "整租每平方米租金中位数",
    ascending=False
)

entire_city_per_sqm

,整租平均每平方米租金,整租每平方米租金中位数
city,,
北京,98.27,93.02
深圳,91.65,85.03
上海,88.16,80.88
广州,53.44,47.13


In [21]:
# 计算有合租样本城市的平均每平方米租金和中位数
shared_city_per_sqm_mean = shared_df.groupby(
    "city"
)["rent_price_per_sqm"].mean().round(2)

shared_city_per_sqm_median = shared_df.groupby(
    "city"
)["rent_price_per_sqm"].median().round(2)

In [22]:
# 把两项统计结果组合成一张表
shared_city_per_sqm = pd.DataFrame({
    "合租平均每平方米租金": shared_city_per_sqm_mean,
    "合租每平方米租金中位数": shared_city_per_sqm_median
})

In [23]:
# 按照每平方米租金中位数从高到低排列
shared_city_per_sqm = shared_city_per_sqm.sort_values(
    "合租每平方米租金中位数",
    ascending=False
)

shared_city_per_sqm

,合租平均每平方米租金,合租每平方米租金中位数
city,,
深圳,171.28,166.25
广州,154.16,153.33
上海,39.91,30.77


In [24]:
# 统一整理四个城市的整租和合租每平方米租金中位数
city_per_sqm_comparison = pd.DataFrame({
    "整租每平方米租金中位数": entire_city_per_sqm_median,
    "合租每平方米租金中位数": shared_city_per_sqm_median
})

In [25]:
# 按照整租每平方米租金中位数从高到低排列
city_per_sqm_comparison = city_per_sqm_comparison.sort_values(
    "整租每平方米租金中位数",
    ascending=False
)

city_per_sqm_comparison

,整租每平方米租金中位数,合租每平方米租金中位数
city,,
北京,93.02,NaN
深圳,85.03,166.25
上海,80.88,30.77
广州,47.13,153.33


## 四、城市房源结构比较

In [26]:
# 复制前面已经统计的城市出租类型数量表
city_rental_structure = city_type_count.copy()

In [27]:
# 计算每个城市的整租占比
city_rental_structure["整租占比（%）"] = (
    city_rental_structure["整租"]
    / city_rental_structure["总计"]
    * 100
).round(2)

# 计算每个城市的合租占比
city_rental_structure["合租占比（%）"] = (
    city_rental_structure["合租"]
    / city_rental_structure["总计"]
    * 100
).round(2)

city_rental_structure

type,合租,整租,总计,整租占比（%）,合租占比（%）
city,,,,,
上海,3,2992,2995,99.90,0.10
北京,0,2992,2992,100.00,0.00
广州,94,2898,2992,96.86,3.14
深圳,317,2682,2999,89.43,10.57
总计,414,11564,11978,96.54,3.46


In [28]:
# 统计每个城市不同卧室数量的整租房源数
entire_bedroom_count = pd.crosstab(   # pd.crosstab()：统计两个分类字段组合后出现的数量
    entire_df["city"],  # 放在行上的字段
    entire_df["bedroom_num"]  # 放在列上的字段
)

entire_bedroom_count

bedroom_num,1.0,2.0,3.0,4.0,5.0,6.0,7.0,8.0,9.0,10.0
city,,,,,,,,,,
上海,629,1200,803,261,83,10,3,0,3,0
北京,816,1366,633,126,40,8,1,1,1,0
广州,852,944,910,144,31,8,4,1,1,0
深圳,1278,496,559,250,67,22,4,4,1,1


In [29]:
# 计算每个城市内部不同卧室数量的整租房源占比
entire_bedroom_percent = pd.crosstab(
    entire_df["city"],
    entire_df["bedroom_num"],
    normalize="index"  # 表示把数量转换为比例
)

# 把小数转换为百分比，并保留两位小数
entire_bedroom_percent = (
    entire_bedroom_percent
    * 100
).round(2)

entire_bedroom_percent

bedroom_num,1.0,2.0,3.0,4.0,5.0,6.0,7.0,8.0,9.0,10.0
city,,,,,,,,,,
上海,21.02,40.11,26.84,8.72,2.77,0.33,0.10,0.00,0.10,0.00
北京,27.27,45.66,21.16,4.21,1.34,0.27,0.03,0.03,0.03,0.00
广州,29.43,32.61,31.43,4.97,1.07,0.28,0.14,0.03,0.03,0.00
深圳,47.65,18.49,20.84,9.32,2.50,0.82,0.15,0.15,0.04,0.04


In [30]:
# 统计每个城市不同卧室数量的合租房源数
shared_bedroom_count = pd.crosstab(
    shared_df["city"],
    shared_df["bedroom_num"]
)

shared_bedroom_count

bedroom_num,1.0,2.0,3.0,4.0,5.0,6.0,7.0,8.0,9.0,13.0,20.0
city,,,,,,,,,,,
上海,2,1,0,0,0,0,0,0,0,0,0
广州,2,0,10,40,28,9,3,0,0,0,2
深圳,4,0,33,163,81,22,7,3,3,1,0


In [31]:
# 计算每个城市内部不同卧室数量的合租房源占比
shared_bedroom_percent = pd.crosstab(
    shared_df["city"],
    shared_df["bedroom_num"],
    normalize="index"
)

# 把小数转换为百分比，并保留两位小数
shared_bedroom_percent = (
    shared_bedroom_percent
    * 100
).round(2)

shared_bedroom_percent

bedroom_num,1.0,2.0,3.0,4.0,5.0,6.0,7.0,8.0,9.0,13.0,20.0
city,,,,,,,,,,,
上海,66.67,33.33,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00
广州,2.13,0.00,10.64,42.55,29.79,9.57,3.19,0.00,0.00,0.00,2.13
深圳,1.26,0.00,10.41,51.42,25.55,6.94,2.21,0.95,0.95,0.32,0.00


## 数据分析总结

### 数据基本情况

- 本次分析使用第3天生成的清洗数据，共11,978条住宅租房记录和26个字段。
- 数据包括11,564条整租房源和414条合租房源。
- 北京有2,992条房源，上海有2,995条，广州有2,992条，深圳有2,999条。
- 北京没有合租样本，上海只有3条合租样本，广州有94条，深圳有317条。
- 合租样本数量较少且城市分布不均衡，因此不能进行可靠的四城市合租市场排名。
- 整租与合租的租金、面积和每平方米租金均分开计算，没有混合分析。

### 主要分析结果

- 北京的整租月租金中位数最高，为6,500元；上海为6,000元，深圳为4,500元，广州为3,100元。
- 上海的整租平均月租金最高，为9,364.25元，但中位数低于北京，说明平均数受到部分高价房源影响。
- 上海的整租面积中位数最大，为90平方米；北京为78平方米，广州为73平方米，深圳为54平方米。
- 北京的整租每平方米租金中位数最高，为93.02元；深圳为85.03元，上海为80.88元，广州为47.13元。
- 上海总租金较高与其整租面积较大有关；深圳典型整租面积最小，但每平方米租金水平较高。
- 北京和上海的整租房源都以两室为主，占比分别为45.66%和40.11%。
- 广州一室、两室和三室整租占比较为接近，户型结构相对均衡。
- 深圳一室整租占47.65%，是四个城市中一室整租占比最高的城市。
- 合租记录中的`bedroom_num`含义不统一，卧室数量表只能描述当前字段的分布，不能解释为城市真实的合租户型占比。
- 合租记录中的`rent_area_num`可能混合出租房间面积和整套住宅面积，因此合租面积和每平方米租金不能用于城市市场排名。